File: Part_II/Data_Structures_and_Algoritms/js/LinkedList.js
const {ListNode} = require('./helpers'); 
// class ListNode {
//     constructor(data = 0, nextNode = null) {
//         this.data = data;
//         this.next = nextNode;
//     }

//     getData() {
//         const nextData = this.next ? this.next.getData() : [];
//         return [this.data, ...nextData]
//     }
// }

function mergeTwoSortedLists(L1, L2) {
    let dummyHead = tail = new ListNode();

    while (L1 && L2) {
        if (L1.data > L2.data) {
            tail.next = L2;
            L2 = L2.next;
        } else {
            tail.next = L1;
            L1 = L1.next;
        }

        tail = tail.next;
    }

    tail.next = L1 || L2;
    return dummyHead.next;
}

const list1 = new ListNode(1, new ListNode(2, new ListNode(5)));
const list2 = new ListNode(3, new ListNode(4, new ListNode(6)));
const mergedList = mergeTwoSortedLists(list1, list2);

// console.log(mergedList.getData()) // -> [1, 2, 3, 4, 5, 6]

function reverseSublist(L, start, finish) {
    let dummyHead = sublistHead = new ListNode(0, L);
    for (let i = 1; i < start; i++) {
        sublistHead = sublistHead.next;
    }

    const sublistIter = sublistHead.next;

    for (let i = 0; i < finish - start; i++) {
        let temp = sublistIter.next;
        sublistIter.next = temp.next;
        temp.next = sublistHead.next;
        sublistHead.next = temp;
    }

    return dummyHead.next;
}

// console.log(reverseSublist(mergedList, 2, 4).getData()); // -> [ 1, 4, 3, 2, 5, 6 ]

function hasCycle(head) {
    function getCycleLen(end) {
        let step = 0;
        let start = end;

        while (true) {
            step += 1;
            start = start.next;

            if (start === end) {
                return step;
            }
        }
    }

    let fast = slow = head;

    while (fast && fast.next && fast.next.next) {
        slow = slow.next;
        fast = fast.next.next;

        if (fast === slow) {
            const cycleLen = getCycleLen(slow);
            let cycleLenAdvancedIter = head;

            for (var i = 0; i < cycleLen; i++) {
                cycleLenAdvancedIter = cycleLenAdvancedIter.next;
            }

            let it = head;
            
            while (it !== cycleLenAdvancedIter) {
                it = it.next;
                cycleLenAdvancedIter = cycleLenAdvancedIter.next;
            }

            return it;
        }
    }

    return null;
}

function hasCycle1(head) {
    let fast = slow = head;

    while (fast && fast.next && fast.next.next) {
        slow = slow.next;
        fast = fast.next.next;

        if (fast === slow) {
            let slow = head;
            
            while (slow !== fast) {
                slow = slow.next;
                fast = fast.next;
            }

            return slow;
        }
    }

    return null;
}

const createNodeArray = (count) => Array(count).fill(1).map((item, i) => new ListNode(i + 1));
const joinNodes = array => array.map((node, i, arr) => {
    node.next = arr[i + 1] || null;
    return node;
});

let cycleList = createNodeArray(6);

cycleList.forEach((node, i, arr) => {
    if (!arr[i + 1]) {
        node.next = arr[2]; // node ListNode(3)
    } else {
        node.next = arr[i + 1];
    }
});

// console.log('hasCycle', hasCycle(cycleList[0]))
// console.log('hasCycle1', hasCycle1(cycleList[0]))

function overlappingNoCycleLists(L1, L2) {
    function length(L) {
        let length = 0;
        
        while (L) {
            L = L.next;
            length += 1;
        }

        return length;
    }

    const L1Len = length(L1);
    const L2Len = length(L2);

    if (L1Len > L2Len) {
        let temp = L2;
        L2 = L1;
        L1 = temp;
    }

    for (var i = 0; i < Math.abs(L1Len - L2Len); i++) {
        L2 = L2.next;
    }

    while (L1 && L2 && L1 !== L2) {
        L1 = L1.next;
        L2 = L2.next;
    }

    return L1;
}

const overlappingL1 = joinNodes(createNodeArray(3));
const overlappingL2 = joinNodes(createNodeArray(6));

overlappingL1.slice(-1)[0].next = overlappingL2[2];

// console.log('overlappingNoCycleLists', overlappingNoCycleLists(overlappingL1[0], overlappingL2[0]))

function overlappingLists(L1, L2) {
    let root1 = hasCycle(L1);
    let root2 = hasCycle(L2);

    if (!root1 && !root2) {
        return overlappingNoCycleLists(L1, L2);
    } else if ((root1 && !root2) || (!root1 && root2)) {
        return null;
    }

    let temp = root2;

    while (true) {
        temp = temp.next;

        if (temp === root1 || temp === root2) {
            break;
        }
    }

    if (temp !== root1) {
        return null;
    }

    function distance(a, b) {
        let dis = 0;
        while (a !== b) {
            a = a.next;
            dis += 1;
        }

        return dis;
    }

    let stem1Length = distance(L1, root1);
    let stem2Length = distance(L2, root2);
    
    if (stem1Length > stem2Length) {
        let temp = L1;
        L1 = L2;
        L2 = temp;
        temp = root1;
        root1 = root2;
        root2 = temp;
    }

    for (var i = 0; i < Math.abs(stem1Length - stem2Length); i++) {
        L2 = L2.next;
    }

    while (L1 !== L2 || L1 !== root1 || L2 !== root2) {
        L1 = L1.next;
        L2 = L2.next;
    }

    return L1 === L2 ? L1 : root1; 
}

// console.log('overlappingLists', overlappingLists(overlappingL1[0], overlappingL2[0]));

function deletionFromList(nodeToDelete) {
    nodeToDelete.data = nodeToDelete.next.data
    nodeToDelete.next = nodeToDelete.next.next
}

function removeKthLast(L, k) {// k-th element from the end
    let dummyHead = new ListNode(0, L);
    let first = dummyHead.next;

    for (let i = 0; i < k; i++) {
        first = first.next;
    }

    let second = dummyHead;

    while (first) {
        first = first.next;
        second = second.next;
    }

    second.next = second.next.next;
    return dummyHead.next;
}
// console.log('removeKthLast 1', removeKthLast(overlappingL2[0], 1).getData());

function removeDuplicates(L) {
    var it = L;

    while (it) {
        var nextDistinct = it.next;

        while (nextDistinct && nextDistinct.data === it.data) {
            nextDistinct = nextDistinct.next;
        }

        it.next = nextDistinct.next;
        it = nextDistinct;
    }

    return L;
}

function cyclicallyRightShiftList(L, k) {
    if (!L) {
        return L;
    }

    var n = 1;
    var tail = L;

    while (tail.next) {
        tail = tail.next;
        n += 1;
    }

    k %= n;

    if (k === 0) {
        return L;
    }

    tail.next = L;
    var stepsToNewHead = n - k;
    var new_tail = L;

    while (stepsToNewHead) {
        stepsToNewHead -= 1;
        new_tail = new_tail.next;
    }

    var new_head = new_tail.next;
    new_tail.next = null;

    return new_head;
}

function evenOddMerge(L) {
    if (!L) {
        return L;
    }

    var evenDummyHead = new ListNode(0);
    var oddDummyHead = new ListNode(0);
    var tails = [evenDummyHead, oddDummyHead];
    var turn = 0;

    while (L) {
        tails[turn].next = L;
        L = L.next;
        tails[turn] = tails[turn].next;
        turn ^= 1;
    }
    
    tails[1].next = null;
    tails[0].next = oddDummyHead.next;
    
    return evenDummyHead.next;
}

function reverseLinkedList(L) {
    let dummyHead = L;
    let dummyTail = null;

    while (dummyHead) {
        const temp = dummyHead.next;
        dummyHead.next = dummyTail;
        dummyTail = dummyHead;

        dummyHead = temp;
    }

    return dummyTail;
}

function isLinkedListAPalindrome(L) {
    // Finds the second half of L
    let slow = fast = L;

    while (fast && fast.next) {
        fast = fast.next.next;
        slow = slow.next;
    }

    // Compare the first half and reversed second half list.
    let firstHalfIter = L;
    let seconfHalfIter = reverseLinkedList(slow);

    while (firstHalfIter && seconfHalfIter) {
        if (firstHalfIter.data != seconfHalfIter.data) {
            return false;
        }

        firstHalfIter = firstHalfIter.next;
        seconfHalfIter = seconfHalfIter.next;
    }

    return true;
}

const list3 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(3, new ListNode(2, new ListNode(1))))));
// console.log('isLinkedListAPalindrome', isLinkedListAPalindrome(list3));

function listPivoting (L, x) {
    let lessHead = lessIter = ListNode();
    let equalHead = equalIter = ListNode();
    let greaterHead = greaterIter = ListNode();

    while (L) {
        if (L.data < x) {
            lessIter.next = L;
            lessIter = lessIter.next;
        } else if (L.data === x) {
            equalIter.next = L;
            equalIter = equalIter.next;
        } else {
            greaterIter.next = L;
            greaterIter = greaterIter.next;
        }

        L = L.next;
    }

    greaterIter.next = null;
    equalIter.next = greaterHead.next;
    lessIter.next = equalHead.next;

    return lessHead.next;
}

function addTwoNumbers(L1, L2) {
    let placeInter = dummyHead = ListNode();
    let curry = 0;

    while (L1 && L2) {
       val = curry + (L1 ? L1.data : 0) + (L2 ? L2.data : 0);
       L1 = L1 ? L1.next : null; 
       L2 = L2 ? L2.next : null;
       placeInter.next = ListNode(val % 10);
       placeInter = placeInter.next;
       curry = Math.floor(val / 10);
    }

    return dummyHead.next;
}

File: Part_II/Data_Structures_and_Algoritms/python/Linked_List.py
from helpers import ListNode

def search_list(L, key):
    while L and L.data != key:
        L = L.next
    
    return L

def insert_after(node, new_node):
    new_node.next = node.next
    node.next = new_node

def delete_after(node):
    node.next = node.next.next

def merge_two_sorted_list(L1, L2):
    dummy_head = tail = ListNode()

    while L1 and L2:
        if L1.data > L2.data:
            tail.next = L2
            L2 = L2.next
        else:
            tail.next = L1
            L1 = L1.next
        tail = tail.next

    tail.next = L1 or L2
    return dummy_head.next

list_1 = ListNode(0, ListNode(1, ListNode(5)))
list_2 = ListNode(2, ListNode(3, ListNode(4)))
merged_list = merge_two_sorted_list(list_1, list_2)

# print(merged_list.getData())

def reverse_sublist(L, start, finish):
    dummy_head = sublist_head = ListNode(0, L)
    for _ in range(1, start):
        sublist_head = sublist_head.next

    sublist_iter = sublist_head.next
    
    for _ in range(finish - start):
        temp = sublist_iter.next
        sublist_iter.next = temp.next
        temp.next = sublist_head.next
        sublist_head.next = temp
    
    return dummy_head.next

list_node = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5, ListNode(6))))))
# print(reverse_sublist(list_node, 2,4).getData())

def has_cycle(head):
    def cycle_len(end):
        start = end
        step = 0

        while True:
            step += 1
            start = start.next
            if start is end:
                return step
    
    fast = slow = head
    while fast and fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

        if slow is fast:
            cycle_len_advanced_iter = head
            for _ in range(cycle_len(slow)):
                cycle_len_advanced_iter = cycle_len_advanced_iter.next

            it = head
            while it is not cycle_len_advanced_iter:
                it = it.next
                cycle_len_advanced_iter = cycle_len_advanced_iter.next
            return it
    return None

## create list

cycleList = [ListNode(i + 1) for i in range(6)]

for i in range(6):
    if i + 1 == len(cycleList):
        cycleList[i].next = cycleList[2]
    else:
        cycleList[i].next = cycleList[i + 1]

## 

# print('has_cycle', vars(has_cycle(cycleList[0])))

def has_cycle_1(head):
    fast = slow = head
    while fast and fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

        if slow is fast:
            slow = head
            while slow is not fast:
                slow = slow.next
                fast = fast.next
            return slow
    return None

# print('has_cycle_1', vars(has_cycle_1(cycleList[0])))

def overlapping_no_cycle_lists(L1, L2):
    def length(L):
        length = 0
        while L:
            L = L.next
            length += 1
        return length

    L1_len = length(L1)
    L2_len = length(L2)

    if L1_len > L2_len:
        temp = L2
        L2 = L1
        L1 = temp

    for _ in range(abs(L2_len - L1_len)):
        L2 = L2.next
    
    while L1 and L2 and L1 is not L2:
        L1 = L1.next
        L2 = L2.next
    return L1

overlappingList1 = [ListNode(i + 1) for i in range(3)]
overlappingList2 = [ListNode(i + 1) for i in range(6)]

for (i, node) in enumerate(overlappingList2):
    if i + 1 != len(overlappingList2):
        overlappingList2[i].next = overlappingList2[i + 1]

for (i, node) in enumerate(overlappingList1):
    if i + 1 == len(overlappingList1):
        overlappingList1[i].next = overlappingList2[3]
    else:
        overlappingList1[i].next = overlappingList1[i + 1]

overlappingList1 = overlappingList1[0]
overlappingList2 = overlappingList2[0]

# print('overlapping_no_cycle_lists', vars(overlapping_no_cycle_lists(overlappingList1, overlappingList2)))

def overlapping_lists(L1, L2):
    root1 = has_cycle(L1)
    root2 = has_cycle(L2)

    if not root1 and not root2:
        return overlapping_no_cycle_lists(L1, L2)
    elif (root1 and not root2) or (not root1 and root2):
        return None

    temp = root2
    while True:
        temp = temp.next
        if temp is root1 or temp is root2:
            break
    
    if temp is not root1:
        return None
    
    def distance(a, b):
        dis = 0
        while a is not b:
            a = a.next
            dis += 1
        return dis
    
    stem1_length = distance(L1, root1)
    stem2_length = distance(L2, root2)

    if stem1_length > stem2_length:
        temp = L1
        L1 = L2
        L2 = temp
        temp = root1
        root1 = root2
        root2 = temp

    for _ in range(abs(stem1_length - stem2_length)):
        L2 = L2.next
    
    while L1 is not L2 and L1 is not root1 and L2 is not root2:
        L1 = L1.next
        L2 = L2.next

    return L1 if L1 is L2 else root1

# print('overlapping_lists', vars(overlapping_lists(overlappingList1, overlappingList2)))

def deletion_from_list(node_to_delete):
    node_to_delete.data = node_to_delete.next.data
    node_to_delete.next = node_to_delete.next.next

def remove_kth_last(L, k):
    dummy_head = ListNode(0, L)
    first = dummy_head.next

    for _ in range(k):
        first = first.next
    
    second = dummy_head
    while first:
        first = first.next
        second = second.next
    
    second.next = second.next.next

    return dummy_head.next

def remove_duplicates(L):
    it = L

    while it:
        next_distinct = it.next

        while next_distinct and next_distinct.data == it.data:
            next_distinct = next_distinct.next
        
        it.next = next_distinct
        it = next_distinct

    return L

def cyclically_right_shift_list(L, k):
    if not L:
        return L

    tail = L
    n = 1

    while tail.next:
        tail = tail.next
        n += 1
    
    k %= n

    if n == 0:
        return L
    
    tail.next = L
    steps_to_new_head = n - k
    new_tail = tail

    while steps_to_new_head:
        new_tail = new_tail.next
        steps_to_new_head -= 1

    new_head = new_tail.next
    new_tail.next = None

    return new_head

def even_odd_merge(L):
    if not L:
        return L
    
    even_dummy_head = ListNode(0)
    odd_dummy_head = ListNode(0)
    tails = [even_dummy_head, odd_dummy_head]
    turn = 0

    while L:
        tails[turn].next = L
        L = L.next
        tails[turn] = tails[turn].next
        turn ^= 1
    
    tails[1].next = None
    tails[0].next = odd_dummy_head.next

    return even_dummy_head.next

def reverse_linked_list(L):
    dummy_head = L
    dummy_tail = None

    while dummy_head:
        temp = dummy_head.next
        dummy_head.next = dummy_tail
        dummy_tail = dummy_head

        dummy_head = temp
    
    return dummy_tail

def is_linked_list_a_palindrome(L):
    # Finds the second half of L
    slow = fast = L

    while fast and fast.next:
        fast, slow = fast.next.next, slow.next

    # Compare the first half and reversed second half list.
    first_half_iter, second_half_iter = L, reverse_linked_list(slow)
    
    while second_half_iter and first_half_iter:
        if second_half_iter.data != first_half_iter.data:
            return False
        second_half_iter = second_half_iter.next
        first_half_iter = first_half_iter.next
    return True

list_node_2 = ListNode(1, ListNode(2, ListNode(3, ListNode(3, ListNode(2, ListNode(1))))))
# print('is_linked_list_a_palindrome', is_linked_list_a_palindrome(list_node_2))

def list_pivoting(L, x):
    less_head = less_iter = ListNode()
    equal_head = equal_iter = ListNode()
    greater_head = greater_iter = ListNode()

    while L:
        if L.data < x:
            less_iter.next = L
            less_iter = less_iter.next
        elif L.data == x:
            equal_head.next = L
            equal_head = equal_head.next
        else:
            greater_head.next = L
            greater_head = greater_head.next

        L = L.next

    greater_iter.next = None
    equal_iter.next = greater_head.next
    less_iter.next = equal_head.next

    return less_head.next


def add_two_numbers(L1, L2):
    place_inter = dummy_head = ListNode()
    carry = 0
    while L1 and L2:
        val = carry + (L1.data if L1 else 0) + (L2.data if L2 else 0)
        L1 = L1.next if L1 else None
        L2 = L2.next if L2 else None
        place_inter.next = ListNode(val % 10)
        carry, place_inter = val // 10, place_inter.next
    return dummy_head.next